src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs

```diff
@@ -180,19 +181,67 @@ protected override async Task OnClientMessageAsync(ConnectionDataMessage connect
             }
-        }
-
-        private Task ProcessClientConnectionAsync(ClientConnectionContext connection)
-        {
+        }
+
+        private async Task ProcessClientConnectionAsync(ClientConnectionContext connection)
+        {
             // Writing from the application to the service
-            var outgoing = ProcessOutgoingMessagesAsync(connection, connection.OutgoingAborted);
+            var transport = ProcessOutgoingMessagesAsync(connection, connection.OutgoingAborted);
 
             // Waiting for the application to shutdown so we can clean up the connection
-            _ = ProcessIncomingMessageAsync(connection);
-
-            // TODO: add more details
-            // Current clean up is inside outgoing task when outgoing task completes
-            return outgoing;
+            var app = ProcessIncomingMessageAsync(connection);
+
+            var task = await Task.WhenAny(app, transport);
+            // anyway remove it from the connection list
+            RemoveClientConnection(connection.ConnectionId);
+
+            // This is the exception from application
+            Exception exception = null;
+            if (task == app)
+            {
+                exception = app.Exception?.GetBaseException();
+
+                // there is no need to write to the transport as application is no longer running
+                Log.WaitingForTransport(Logger);
+
+                // app task completes connection.Transport.Output, which will completes connection.Application.Input and ends the transport 
+                // Transports are written by us and are well behaved, wait for them to drain
+                connection.CancelOutgoing(_closeTimeOutMilliseconds);
+                await transport;
+            }
+            else
+            {
+                // transport task ends first, no data will be dispatched out
+                Log.WaitingForApplication(Logger);
+
+                // Wait on the application task to complete
+                connection.CancelApplication(_closeTimeOutMilliseconds);
+                try
+                {
+                    await app;
+                }
+                catch (Exception e)
+                {
+                    exception = e;
+                }
+            }
+
+            if (exception != null)
+            {
+                Log.ApplicationTaskFailed(Logger, exception);
+            }
+
+            // If we aren't already aborted, we send the abort message to the service
+            if (connection.AbortOnClose)
+            {
+                // Inform the Service that we will remove the client because SignalR told us it is disconnected.
+                var serviceMessage =
+                    new CloseConnectionMessage(connection.ConnectionId, errorMessage: exception?.Message);
+                await WriteAsync(serviceMessage);
+                Log.CloseConnection(Logger, connection.ConnectionId);
+            }
+
```

Comment:
There is a `connection.OnCompleted();` that has been called at the end of the `ProcessOutgoingMessagesAsync`.

I suppose that can be moved to here since here is the real end of a connection after refactor.

Reply:
yes nice catch